subsys/bluetooth/host/att.c

```diff
@@ -2948,6 +2968,49 @@ static int bt_att_accept(struct bt_conn *conn, struct bt_l2cap_chan **ch)
 BT_L2CAP_CHANNEL_DEFINE(att_fixed_chan, BT_L2CAP_CID_ATT, bt_att_accept, NULL);
 
 #if defined(CONFIG_BT_EATT)
+static k_timeout_t credit_based_connection_delay(struct bt_conn *conn)
+{
+	/*
+	 * 5.3 Vol 3, Part G, Section 5.4 L2CAP COLLISION MITIGATION
+	 * ... In this situation, the Central may retry
+	 * immediately but the Peripheral shall wait a minimum of 100 ms before retrying;
+	 * on LE connections, the Peripheral shall wait at least 2 *
+	 * (connPeripheralLatency + 1) * connInterval if that is longer.
+	 */
+
+	if (conn->role == BT_CONN_ROLE_CENTRAL) {
+		return K_NO_WAIT;
+	} else {
+		return K_MSEC(MAX(100, 2 * (conn->le.latency + 1) * conn->le.interval));
+	}
+}
+
+int att_schedule_eatt_connect(struct bt_conn *conn, struct att_connection_work_context context)
+{
+	struct bt_att *att = att_get(conn);
+
+	att->connection_work_context = context;
+	return k_work_schedule(&att->connection_work, credit_based_connection_delay(conn));
+}
+
+void ecred_connect_cb(struct bt_conn *conn, uint16_t result, uint8_t attempted_to_connect,
+		      uint8_t succeded_to_connect)
+{
+	static uint16_t previous_result;
```

Comment:
hmm this will be common for all connections,  maybe extend callback to pass some extra context (or user data)? (but that means it would have to be per channel callback, not global)

Reply:
Yes, this cannot be global. Will make it per-connection.